node/src/p2p.rs

```diff
@@ -66,37 +44,100 @@ pub enum P2pError {
     WorkerDied,
 }
 
-type Result<T, E = P2pError> = std::result::Result<T, E>;
+#[derive(Debug)]
+pub struct P2p {
+    cmd_tx: flume::Sender<P2pCmd>,
+}
 
-struct Worker {
-    swarm: Swarm<Behaviour>,
-    header_sub_topic_hash: TopicHash,
-    cmd_rx: flume::Receiver<P2pCmd>,
+pub struct P2pArgs {
+    pub transport: Boxed<(PeerId, StreamMuxerBox)>,
+    pub network_id: String,
+    pub local_keypair: Keypair,
+    pub bootstrap_peers: Vec<Multiaddr>,
+    pub listen_on: Vec<Multiaddr>,
 }
 
 #[derive(Debug)]
-enum P2pCmd {
+pub enum P2pCmd {
     NetworkInfo {
         respond_to: oneshot::Sender<NetworkInfo>,
     },
 }
 
-#[allow(unused)]
-pub enum P2pEvent {}
+#[async_trait]
+impl Service for P2p {
+    type Command = P2pCmd;
+    type Args = P2pArgs;
+    type Error = P2pError;
+
+    async fn start(args: P2pArgs) -> Result<Self, P2pError> {
+        let (cmd_tx, cmd_rx) = flume::bounded(16);
+        let mut worker = Worker::new(args, cmd_rx)?;
 
-impl P2p {
-    pub fn new(config: P2pConfig) -> Result<P2p> {
-        let local_peer_id = PeerId::from(config.local_keypair.public());
+        spawn(async move {
+            worker.run().await;
+        });
+
+        Ok(P2p { cmd_tx })
+    }
+
+    async fn stop(&self) -> Result<()> {
+        todo!()
+    }
+
+    async fn send_command(&self, cmd: P2pCmd) -> Result<()> {
+        self.cmd_tx
+            .send_async(cmd)
+            .await
+            .map_err(|_| P2pError::WorkerDied)
+    }
+}
+
+#[async_trait]
+pub trait P2pService: Service<Args = P2pArgs, Command = P2pCmd, Error = P2pError> {
+    async fn network_info(&self) -> Result<NetworkInfo> {
+        let (tx, rx) = oneshot::channel();
+
+        self.send_command(P2pCmd::NetworkInfo { respond_to: tx })
+            .await?;
+        rx.await
+            .map_err(|_| P2pError::CmdCancelled("NetworkInfo".to_owned()))
```

Comment:
Shouldn't this be `WorkerDied`? I don't see how a command can be cancelled

Reply:
removed cmdcancelled entirely